src/client/gemini_client.rs

```diff
@@ -0,0 +1,308 @@
+use bytes::Bytes;
+use nu_protocol::ShellError;
+use reqwest::Response;
+use serde::{Deserialize, Serialize};
+use tokio::{select, time::sleep, time::Duration};
+
+pub struct GeminiClient {
+    api_key: String,
+    max_tokens: usize,
+}
+
+// While Gemini does not have a per request limit, the per minute token limit is used here
+const MAX_FREE_TIER_TOKENS: usize = 1000000;
+
+// According to the Gemini API docs: A token is equivalent to about 4 characters for Gemini models
+const CHARS_PER_TOKEN: usize = 4;
+
+const MAX_EMBEDDING_DIMENSION: u32 = 768;
+
+// At most 100 requests can be in one batch
+const MAX_BATCH_SIZE: usize = 100;
+
+impl GeminiClient {
+    pub fn new(api_key: String, max_tokens: impl Into<Option<usize>>) -> Self {
+        let max_tokens = match max_tokens.into() {
```

Comment:
Another not saying we should do I think we could actually something like `let max_tokens = max_tokens.into().unwrap_or(MAX_FREE_TIER_TOKENS)`.